crates/ibc/src/core/ics02_client/handler/create_client.rs

```diff
@@ -37,14 +37,20 @@ where
 
     let client_type = client_state.client_type();
 
-    let _client_id = ClientId::new(client_type, id_counter).map_err(|e| {
+    let client_id = ClientId::new(client_type, id_counter).map_err(|e| {
         ClientError::ClientIdentifierConstructor {
             client_type: client_state.client_type(),
             counter: id_counter,
             validation_error: e,
         }
     })?;
 
+    let stored_client_state = ctx.client_state(&client_id);
+    match stored_client_state {
+        Err(ContextError::ClientError(ClientError::ClientStateNotFound { client_id: _ })) => (),
```

Comment:
I wouldn't rely on users actually returning `ClientStateNotFound`. As we do [elsewhere](https://github.com/cosmos/ibc-rs/blob/caff666ced4c95ff04f59c798f5e4870e0dcd44a/crates/ibc/src/core/context/timeout.rs#L78), we should just check `is_err()` here. I would wait for my [suggested error system redesign](https://github.com/cosmos/ibc-rs/issues/270#issuecomment-1470734150) to be implemented before we match on specific errors returned from users

Reply:
I see, but I'd like to point out that relying on the `is_err()` still leaves the validation step dependent on the users. wouldn't be much better. If `client_state()` fails due to any other reasons such as "broken storage" or "unable to fetch from storage", etc... the validation will wrongly pass, which is why I created issue #607. 
I think, this is a significant one needing to be addressed even before we begin to refactor our errors.

Reply:
Note that using `is_err()` achieves the exact [same behavior from ibc-go](https://github.com/cosmos/ibc-go/blob/ae27fa5960d523e4476776308ecd0b4cd3377671/modules/core/04-channel/keeper/packet.go#L210). Being more specific about which error occurred would be good for sure, but I consider using `is_err()` as at least following the same standard as ibc-go.

Reply:
Yes, though IBC-go's situation is better than ours. The `nil` is handled internally, users' preferences in error picking do not affect that. Plus, the storage is in a more known and clearer position with respect to the entire IBC implementation, which makes it less susceptible to errors, but it still suffers from this problem.
np, I went with the `is_err` approach for this PR ([c9eba3e](https://github.com/cosmos/ibc-rs/pull/605/commits/c9eba3ec7bc0c051dd8a742526d61edb8d99726f)) Though, the discussed issue remains there.